File: frontend/monaco_editor.js
/**
 * monaco_editor.js
 * ----------------
 * Vue 3 component wrapping Monaco Editor.
 *
 * Web Worker Architecture
 * -----------------------
 * Monaco offloads language processing (tokenization, bracket matching, diff
 * computation) to Web Workers. Because browsers require workers to come from
 * the same origin as the page, we cannot point directly to CDN worker URLs.
 *
 * Solution: before the AMD loader is invoked, we set `window.MonacoEnvironment`
 * to redirect worker URLs to our FastAPI static endpoint
 * (`/monaco-workers/<type>.worker.js`). Each proxy script at that URL calls
 * `importScripts()` to fetch the real Monaco worker bundle from the CDN inside
 * the worker scope — which IS permitted by the browser.
 *
 * Worker URL routing:
 *   typescript / javascript  →  /monaco-workers/ts.worker.js
 *   json                     →  /monaco-workers/json.worker.js
 *   css / less / scss        →  /monaco-workers/css.worker.js
 *   everything else          →  /monaco-workers/editor.worker.js  (general)
 *
 * Diff Editor
 * -----------
 * When `diff_mode=True`, Python passes both `value` (modified/AI side) and
 * `original_value` (legacy side). Monaco computes line-level diff natively
 * client-side inside the diff worker — no Python involvement after mount.
 */

const MONACO_CDN   = 'https://cdn.jsdelivr.net/npm/monaco-editor@0.52.2/min/vs';
const MONACO_LOADER = `${MONACO_CDN}/loader.js`;
// FastAPI serves worker proxies at this base (port 8000)
const WORKERS_BASE  = 'http://localhost:8000/monaco-workers';

/** Inject MonacoEnvironment BEFORE loading editor.main so workers are routed. */
function _installWorkerEnv() {
    if (window.__monacoWorkerEnvInstalled) return;
    window.__monacoWorkerEnvInstalled = true;
    window.MonacoEnvironment = {
        getWorkerUrl(_moduleId, label) {
            if (label === 'typescript' || label === 'javascript') {
                return `${WORKERS_BASE}/ts.worker.js`;
            }
            if (label === 'json') {
                return `${WORKERS_BASE}/json.worker.js`;
            }
            if (label === 'css' || label === 'less' || label === 'scss') {
                return `${WORKERS_BASE}/css.worker.js`;
            }
            // Default: general editor worker (handles Python, C, C++, plaintext, etc.)
            return `${WORKERS_BASE}/editor.worker.js`;
        }
    };
}

/** Singleton promise — only one <script> tag is ever injected. */
function _loadMonaco() {
    if (window.__monacoLoadPromise) return window.__monacoLoadPromise;
    _installWorkerEnv();   // Must be set before AMD loader fires
    window.__monacoLoadPromise = new Promise((resolve, reject) => {
        const script = document.createElement('script');
        script.src = MONACO_LOADER;
        script.onerror = () => reject(new Error(`Failed to load Monaco loader from ${MONACO_LOADER}`));
        script.onload = () => {
            require.config({ paths: { vs: MONACO_CDN } });
            require(['vs/editor/editor.main'], resolve, reject);
        };
        document.head.appendChild(script);
    });
    return window.__monacoLoadPromise;
}

export default {
    template: `
        <div :style="{ width: '100%', height: height, overflow: 'hidden', position: 'relative' }" ref="container">
            <div ref="editorNode" style="position: absolute; top: 0; left: 0; right: 0; bottom: 0;"></div>
        </div>
    `,
    props: {
        value:          { type: String,  default: '' },
        language:       { type: String,  default: 'python' },
        readonly:       { type: Boolean, default: false },
        diff_mode:      { type: Boolean, default: false },
        original_value: { type: String,  default: '' },
        primary_line:   { type: Number,  default: 0 },
        height:         { type: String,  default: '600px' },
        debounce_delay: { type: Number,  default: 1000 },
    },
    emits: ['change', 'save'],

    data() {
        return {
            _lastEmittedValue: this.value || '',
        };
    },

    mounted() {
        this._debounceTimer = null;
        this._lastEmittedValue = this.value || '';
        _installWorkerEnv();

        if (window.monaco) {
            this.initEditor();
            return;
        }
        _loadMonaco().then(() => this.initEditor()).catch(err => {
            console.error('[MonacoEditor] Failed to load Monaco:', err);
        });
    },

    methods: {
        getCurrentValue() {
            if (!this._editor) return this._lastEmittedValue;
            if (this.diff_mode) {
                return this._modifiedModel ? this._modifiedModel.getValue() : '';
            }
            return this._editor.getValue ? this._editor.getValue() : '';
        },

        /**
         * Flush pending changes to Python (Canonical state).
         * @param {boolean} isManualSave - Whether triggered explicitly by user save action.
         */
        flushChange(isManualSave = false) {
            if (this._debounceTimer) {
                clearTimeout(this._debounceTimer);
                this._debounceTimer = null;
            }

            const current = this.getCurrentValue();
            if (current !== this._lastEmittedValue || isManualSave) {
                this._lastEmittedValue = current;
                this.$emit('change', current);
                if (isManualSave) {
                    this.$emit('save', current);
                }
            }
        },

        /**
         * Schedule local debounced sync (1000ms delay).
         */
        scheduleDebouncedChange() {
            if (this.readonly) return;
            if (this._debounceTimer) {
                clearTimeout(this._debounceTimer);
            }
            const delay = typeof this.debounce_delay === 'number' ? this.debounce_delay : 1000;
            this._debounceTimer = setTimeout(() => {
                this.flushChange(false);
            }, delay);
        },

        initEditor() {
            const node = this.$refs.editorNode;
            if (!node) return;

            const commonOpts = {
                theme: 'vs-dark',
                automaticLayout: false,
                scrollBeyondLastLine: false,
                minimap: { enabled: false },
                fontSize: 13,
                lineHeight: 20,
            };

            if (this.diff_mode) {
                // -------------------------------------------------------
                // DIFF EDITOR
                // -------------------------------------------------------
                this._editor = monaco.editor.createDiffEditor(node, {
                    ...commonOpts,
                    readOnly: this.readonly,
                    renderSideBySide: true,
                    ignoreTrimWhitespace: false,
                    renderIndicators: true,
                    originalEditable: false,
                });

                this._originalModel = monaco.editor.createModel(
                    this.original_value || '', this.language || 'python'
                );
                this._modifiedModel = monaco.editor.createModel(
                    this.value || '', this.language || 'python'
                );

                this._editor.setModel({
                    original: this._originalModel,
                    modified: this._modifiedModel,
                });

                // Debounced content change listener
                this._modifiedModel.onDidChangeContent(() => {
                    this.scheduleDebouncedChange();
                });

                // Flush on editor blur (Ephemeral -> Canonical)
                const modifiedEditor = this._editor.getModifiedEditor();
                if (modifiedEditor) {
                    modifiedEditor.onDidBlurEditorText(() => {
                        this.flushChange(false);
                    });

                    // Ctrl+S / Cmd+S manual save handler
                    modifiedEditor.addCommand(monaco.KeyMod.CtrlCmd | monaco.KeyCode.KeyS, () => {
                        this.flushChange(true);
                    });
                }

            } else {
                // -------------------------------------------------------
                // PLAIN EDITOR
                // -------------------------------------------------------
                this._editor = monaco.editor.create(node, {
                    ...commonOpts,
                    value: this.value || '',
                    language: this.language || 'python',
                    readOnly: this.readonly,
                });

                // Debounced content change listener
                this._editor.onDidChangeModelContent(() => {
                    this.scheduleDebouncedChange();
                });

                // Flush on editor blur
                this._editor.onDidBlurEditorText(() => {
                    this.flushChange(false);
                });

                // Ctrl+S / Cmd+S manual save handler
                this._editor.addCommand(monaco.KeyMod.CtrlCmd | monaco.KeyCode.KeyS, () => {
                    this.flushChange(true);
                });
            }

            // Force layout passes
            const forceLayout = () => { if (this._editor) this._editor.layout(); };
            requestAnimationFrame(() => {
                forceLayout();
                requestAnimationFrame(forceLayout);
            });
            setTimeout(forceLayout, 200);

            // ResizeObserver
            this._resizeObserver = new ResizeObserver(entries => {
                if (!entries.length) return;
                const { width, height } = entries[0].contentRect;
                const w = Math.round(width), h = Math.round(height);
                if (this._lastW !== w || this._lastH !== h) {
                    this._lastW = w; this._lastH = h;
                    requestAnimationFrame(() => { if (this._editor) this._editor.layout(); });
                }
            });
            this._resizeObserver.observe(this.$refs.container);

            this._applyDecorations();
        },

        _applyDecorations() {
            if (!this.primary_line || this.primary_line <= 0 || !this._editor) return;
            const target = this.diff_mode
                ? this._editor.getModifiedEditor()
                : this._editor;
            this._decorations = target.createDecorationsCollection([{
                range: new monaco.Range(this.primary_line, 1, this.primary_line, 1),
                options: { isWholeLine: true, className: 'monaco-error-primary' }
            }]);
            target.revealLineInCenter(this.primary_line);
        },

        // Client-side methods exposed for Python run_method
        flush() {
            this.flushChange(false);
        },
        save() {
            this.flushChange(true);
        },
    },

    watch: {
        /** Update the modified (AI) model content when Python pushes a new canonical value. */
        value(newVal) {
            if (!this._editor) return;
            const safe = newVal || '';
            const current = this.getCurrentValue();
            if (safe !== current) {
                this._lastEmittedValue = safe;
                if (this.diff_mode) {
                    if (this._modifiedModel) this._modifiedModel.setValue(safe);
                } else {
                    this._editor.setValue(safe);
                }
            }
        },

        /** Update the original (legacy) model content when Python changes it. */
        original_value(newVal) {
            if (!this._editor || !this.diff_mode) return;
            const safe = newVal || '';
            if (this._originalModel && safe !== this._originalModel.getValue()) {
                this._originalModel.setValue(safe);
            }
        },

        /** Re-apply decorations when the error line changes (e.g., after retest). */
        primary_line() {
            if (this._decorations) {
                this._decorations.clear();
                this._decorations = null;
            }
            this._applyDecorations();
        },
    },

    beforeUnmount() {
        if (this._debounceTimer) {
            this.flushChange(false);
        }
        if (this._resizeObserver) this._resizeObserver.disconnect();
        if (this._editor) {
            if (this.diff_mode) {
                this._originalModel?.dispose();
                this._modifiedModel?.dispose();
            } else {
                this._editor.getModel()?.dispose();
            }
            this._editor.dispose();
        }
        this._editor = this._originalModel = this._modifiedModel = null;
    },
};

File: frontend/monaco_editor.py
"""
frontend/monaco_editor.py
--------------------------
NiceGUI 3.x custom Element that wraps Monaco Editor via a local Vue 3
component (monaco_editor.js in the same directory).

Web Worker Architecture
-----------------------
Monaco offloads CPU-intensive language processing (diff computation, syntax
tokenization, bracket matching) to dedicated Web Workers.  Because browsers
enforce a same-origin policy for workers, we cannot point directly at CDN
worker URLs.

Solution implemented in monaco_editor.js:
  1. ``window.MonacoEnvironment.getWorkerUrl()`` is installed before AMD
     ``require(['vs/editor/editor.main'])`` fires.
  2. The function returns a FastAPI-served URL:
     ``http://localhost:8000/monaco-workers/<type>.worker.js``
  3. Each proxy file at that URL calls ``importScripts()`` to fetch the real
     Monaco worker bundle from the CDN — importScripts is permitted inside
     a Worker context regardless of origin.

FastAPI serves the proxy scripts via:
    app.mount("/monaco-workers", StaticFiles(directory="frontend/monaco_workers"))

Diff Editor
-----------
When ``diff_mode=True``, Monaco creates a ``DiffEditor`` with two models:
  * **original** model  ← ``original_value``  (legacy / left pane)
  * **modified** model  ← ``value``            (AI patch / right pane)

Monaco computes the unified diff natively inside the diff worker — no
Python calculation needed.  Watchers on both ``value`` and ``original_value``
props push updates to the client-side models without re-mounting the editor.

Usage
-----
    from frontend.monaco_editor import MonacoEditor

    # Read-only viewer
    MonacoEditor(value=code, language="python", readonly=True)
        .classes("w-full").style("height:520px;")

    # Monaco diff editor (side-by-side, read-only)
    MonacoEditor(
        value=ai_code,
        original_value=legacy_code,
        language="python",
        readonly=True,
        diff_mode=True,
        primary_line=42,        # optional: scroll & highlight error line
    ).classes("w-full").style("height:600px;")

    # Writable editor — receive changes in Python
    def handle_change(new_text: str):
        state.edit_buffer[file_id] = new_text

    MonacoEditor(
        value=current_draft,
        original_value=original_draft,
        language="python",
        readonly=False,
        diff_mode=True,
        on_change=handle_change,
    ).classes("w-full").style("height:480px;")
"""

from __future__ import annotations

from typing import Callable, Optional
from nicegui.element import Element

# ---------------------------------------------------------------------------
# Language normalisation
# ---------------------------------------------------------------------------

_LANG_MAP: dict[str, str] = {
    "python":     "python",
    "r":          "r",
    "c":          "c",
    "cpp":        "cpp",
    "js":         "javascript",
    "javascript": "javascript",
    "ts":         "typescript",
    "typescript": "typescript",
    "json":       "json",
    "css":        "css",
    "html":       "html",
}


def _monaco_lang(language: str) -> str:
    """Map internal language codes to Monaco language identifiers."""
    return _LANG_MAP.get((language or "").lower(), "plaintext")


# ---------------------------------------------------------------------------
# MonacoEditor NiceGUI Element
# ---------------------------------------------------------------------------

class MonacoEditor(Element, component="monaco_editor.js"):
    """
    A Monaco Editor embedded in NiceGUI via a Vue 3 component.

    Language-specific Web Workers are routed through FastAPI's
    ``/monaco-workers/`` static endpoint so they satisfy the browser's
    same-origin constraint for workers.

    Parameters
    ----------
    value : str
        Source code content (the "modified" / AI side in diff mode).
    language : str
        Syntax highlighting language. Accepts ``'python'``, ``'r'``,
        ``'c'``, ``'cpp'``, ``'javascript'``, ``'typescript'``, ``'json'``,
        ``'css'``, ``'html'``, etc.
    readonly : bool
        When ``True`` the editor is non-editable.
    diff_mode : bool
        When ``True`` renders a side-by-side Monaco diff editor.
        ``original_value`` is used as the left (legacy) side.
        Monaco computes the diff natively client-side.
    original_value : str
        The "original" (left/legacy) content shown in diff mode.
    primary_line : int | None
        If provided, the editor scrolls to this line and applies an
        amber highlight decoration (used to surface sandbox error lines).
    height : str
        CSS height for the editor container (default ``'600px'``).
    debounce_delay : int
        Local millisecond delay for debouncing text changes before pushing
        canonical updates back to Python over WebSocket (default ``1000``).
    on_change : callable | None
        Python callback invoked when content changes after the 1000ms debounce
        or on editor blur. Receives the updated code ``str``.
    on_save : callable | None
        Python callback invoked when user triggers a manual save (Ctrl+S / Cmd+S).
    """

    def __init__(
        self,
        value: str = "",
        language: str = "python",
        readonly: bool = True,
        diff_mode: bool = False,
        original_value: str = "",
        primary_line: Optional[int] = None,
        height: str = "600px",
        debounce_delay: int = 1000,
        on_change: Optional[Callable[[str], None]] = None,
        on_save: Optional[Callable[[str], None]] = None,
    ) -> None:
        super().__init__()

        lang = _monaco_lang(language)

        # These map directly to the Vue component's `props`
        self._props["value"]          = value
        self._props["language"]       = lang
        self._props["readonly"]       = readonly
        self._props["diff_mode"]      = diff_mode
        self._props["original_value"] = original_value
        self._props["primary_line"]   = primary_line or 0
        self._props["height"]         = height
        self._props["debounce_delay"] = debounce_delay

        if on_change:
            # Vue emits: this.$emit('change', editor.getValue())
            # NiceGUI delivers e.args as [value] or dict
            self.on("change", lambda e: on_change(
                e.args[0] if isinstance(e.args, (list, tuple)) and e.args else (e.args if isinstance(e.args, str) else "")
            ))

        if on_save:
            # Vue emits: this.$emit('save', editor.getValue())
            self.on("save", lambda e: on_save(
                e.args[0] if isinstance(e.args, (list, tuple)) and e.args else (e.args if isinstance(e.args, str) else "")
            ))

    # ------------------------------------------------------------------
    # Programmatic update helpers — push property updates to the client
    # without destroying and re-creating the editor instance.
    # ------------------------------------------------------------------

    def update_value(self, value: str) -> "MonacoEditor":
        """Replace the modified (AI) side content from Python."""
        self._props["value"] = value
        self.update()
        return self

    def update_original_value(self, original_value: str) -> "MonacoEditor":
        """Replace the original (legacy) side content in diff mode."""
        self._props["original_value"] = original_value
        self.update()
        return self

    def update_language(self, language: str) -> "MonacoEditor":
        """Switch the syntax highlighting language."""
        self._props["language"] = _monaco_lang(language)
        self.update()
        return self

    def update_readonly(self, readonly: bool) -> "MonacoEditor":
        """Toggle editor read-only state."""
        self._props["readonly"] = readonly
        self.update()
        return self

    async def flush(self) -> None:
        """Trigger immediate client-side flush of pending debounced changes."""
        self.run_method("flush")

    async def save(self) -> None:
        """Trigger client-side manual save action."""
        self.run_method("save")
